Features/state and props access within setstate

Suzii opened the pull request.

Suzii committed:
Refactor callback function in walking algorithm

Suzii committed:
Fix .gitignore and remove generated files from repo

Suzii committed:
Make rule detect this.state|props within setState(...)

Previous implementation only detected setState(...) with object literal
passed. However, the real issue lies in the fact that 'this.state' nor
'this.props' must not be accessed within setState(...) even for functional
setState.

PetrSvirak reviewed (changes requested):
It is all great. Really nice refactoring (one might even say beautification) and finishing of the rule. Thanks :+1: 

There are but two things: 
* It would seem a bit better to highlight `this.state.` or `this.props.` instead of whole `updater` function – it would be more straightforward, one would see all occurrences and, for longer states, it would not red half of the screen :stuck_out_tongue_closed_eyes: 
* Also, please, there should be a test for multiple usages of `this.state.` or `this.props.` or at least their combination. Could you please add one for all options?

PetrSvirak committed:
Update and unify failure messages

PetrSvirak committed:
Highlight only this.props and this.state used within setState.

Seems a bit more accurate to only highlight the potential cause of problem.
I also added some corner cases tests and, as a result, fixed a false negative
unreported when first argument of setState was an object in parentheses.

PetrSvirak committed:
Add TS code style and reformat existing files

Suzii committed:
Make 'removeParentheses' function recursive and add tests

Suzii reviewed (commented):
Thanks for the changes, I think it is awesome now. I left some inline comments which you might consider, but generally it is good to go.

PetrSvirak commented on src/utils/syntaxWalkerUtils.ts:
```diff
@@ -25,3 +26,10 @@ export function getFirstSetStateAncestor(node: ts.Node): ts.CallExpression {
     return getFirstSetStateAncestor(node.parent);
 }
+
+export function removeParentheses(node: ts.Node){
```
http://www.commitstrip.com/en/2016/09/30/commit-or-die-trying/
